### PR TITLE
test typescriptification - spec/integ

### PR DESCRIPTION
**Pull request by kerryarchibald**
The final instalment of test typescriptification 🥳

Fixes https://github.com/matrix-org/matrix-js-sdk/issues/2353

## Checklist

* [ ] Tests written for new code (and old code if feasible)
* [ ] Linter and other CI checks pass
* [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

**Review by turt2live — COMMENTED** [review]
clearing code review while draft

File: spec/integ/devicelist-integ.spec.ts [hunk 1]
```diff
@@ -290,8 +290,9 @@ describe("DeviceList management:", function() {
             aliceTestClient.client.cryptoStore.getEndToEndDeviceData(null, (data) => {
                 const bobStat = data.trackingStatus['@bob:xyz'];
 
+                // Alice should be tracking bob's device list
                 expect(bobStat).toBeGreaterThan(
-                    0, "Alice should be tracking bob's device list",
+                    0,
```
Comment (duxovni) [inline spec/integ/devicelist-integ.spec.ts:295]: May as well just make this a single line

File: spec/integ/devicelist-integ.spec.ts [hunk 2]
```diff
@@ -326,8 +327,9 @@ describe("DeviceList management:", function() {
             aliceTestClient.client.cryptoStore.getEndToEndDeviceData(null, (data) => {
                 const bobStat = data.trackingStatus['@bob:xyz'];
 
+                // Alice should have marked bob's device list as untracked
                 expect(bobStat).toEqual(
-                    0, "Alice should have marked bob's device list as untracked",
+                    0,
```
Comment (duxovni) [inline spec/integ/devicelist-integ.spec.ts:332]: Likewise

File: spec/integ/devicelist-integ.spec.ts [hunk 3]
```diff
@@ -362,8 +364,9 @@ describe("DeviceList management:", function() {
...
```
Comment (duxovni) [inline spec/integ/devicelist-integ.spec.ts:369]: ^

File: spec/integ/devicelist-integ.spec.ts [hunk 4]
```diff
@@ -378,13 +381,15 @@ describe("DeviceList management:", function() {
...
 
+                // @ts-ignore accessing private property
                 anotherTestClient.client.cryptoStore.getEndToEndDeviceData(null, (data) => {
-                    const bobStat = data.trackingStatus['@bob:xyz'];
+                    const bobStat = data!.trackingStatus['@bob:xyz'];
 
+                    // Alice should have marked bob's device list as untracked
                     expect(bobStat).toEqual(
-                        0, "Alice should have marked bob's device list as untracked",
+                        0,
```
Comment (duxovni) [inline spec/integ/devicelist-integ.spec.ts:392]: ^

File: spec/integ/matrix-client-event-emitter.spec.ts [hunk 5]
```diff
@@ -1,19 +1,53 @@
...
         httpBackend.when("GET", "/pushrules").respond(200, {});
         httpBackend.when("POST", "/filter").respond(200, { filter_id: "a filter id" });
+
+        return [client, httpBackend];
+    };
+
+    let [client, httpBackend] = setupTests();
+
+    beforeEach(function() {
+        [client, httpBackend] = setupTests();
```
Comment (duxovni) [inline spec/integ/matrix-client-event-emitter.spec.ts:50]: Is there a way of doing this that doesn't leave a running httpBackend lying around?

File: spec/integ/matrix-client-methods.spec.ts [hunk 6]
```diff
@@ -13,35 +13,48 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-        httpBackend = testClient.httpBackend;
-        client = testClient.client;
+
+        return [testClient.client, testClient.httpBackend, store];
+    };
+
+    let [client, httpBackend, store] = setupTests();
+
+    beforeEach(function() {
+        [client, httpBackend, store] = setupTests();
```
Comment (duxovni) [inline spec/integ/matrix-client-methods.spec.ts:57]: Here as well

File: spec/integ/matrix-client-methods.spec.ts [hunk 7]
```diff
@@ -56,21 +69,23 @@ describe("MatrixClient", function() {
                 "POST", "/_matrix/media/r0/upload",
             ).check(function(req) {
                 expect(req.rawData).toEqual(buf);
-                expect(req.queryParams.filename).toEqual("hi.txt");
-                if (!(req.queryParams.access_token == accessToken ||
+                expect(req.queryParams?.filename).toEqual("hi.txt");
+                if (!(req.queryParams?.access_token == accessToken ||
                         req.headers["Authorization"] == "Bearer " + accessToken)) {
```
Comment (duxovni) [inline spec/integ/matrix-client-methods.spec.ts:74]: These should be `===` right?

File: spec/integ/matrix-client-opts.spec.ts [hunk 8]
```diff
@@ -99,7 +101,7 @@ describe("MatrixClient opts", function() {
             ];
             client.on("event", function(event) {
                 expect(expectedEventTypes.indexOf(event.getType())).not.toEqual(
-                    -1, "Recv unexpected event type: " + event.getType(),
+                    -1,
                 );
```
Comment (duxovni) [inline spec/integ/matrix-client-opts.spec.ts:105]: This can also be a one-liner now

File: spec/integ/matrix-client-opts.spec.ts [hunk 9]
```diff
@@ -118,7 +120,7 @@ describe("MatrixClient opts", function() {
                 utils.syncPromise(client),
             ]);
             expect(expectedEventTypes.length).toEqual(
-                0, "Expected to see event types: " + expectedEventTypes,
+                0,
             );
```
Comment (duxovni) [inline spec/integ/matrix-client-opts.spec.ts:124]: And here

File: spec/integ/matrix-client-retrying.spec.ts [hunk 10]
```diff
@@ -14,33 +14,38 @@ See the License for the specific language governing permissions and
...
         client.store.storeRoom(room);
+
+        return [client, httpBackend, room];
+    };
+
+    let [client, httpBackend, room] = setupTests();
+
+    beforeEach(function() {
+        [client, httpBackend, room] = setupTests();
     });
```
Comment (duxovni) [inline spec/integ/matrix-client-retrying.spec.ts:49]: Loose httpBackend here too

File: spec/integ/matrix-client-room-timeline.spec.ts [hunk 11]
```diff
@@ -119,6 +125,13 @@ describe("MatrixClient room timelines", function() {
         });
         client.startClient();
 
+        return [client, httpBackend];
+    };
+
+    let [client, httpBackend] = setupTestClient();
+
+    beforeEach(async function() {
+        [client, httpBackend] = setupTestClient();
```
Comment (duxovni) [inline spec/integ/matrix-client-room-timeline.spec.ts:134]: And here

File: spec/integ/matrix-client-syncing.spec.ts [hunk 12]
```diff
@@ -45,13 +47,20 @@ describe("MatrixClient syncing", () => {
...
         httpBackend!.when("GET", "/versions").respond(200, {});
         httpBackend!.when("GET", "/pushrules").respond(200, {});
         httpBackend!.when("POST", "/filter").respond(200, { filter_id: "a filter id" });
+        return [client, httpBackend];
+    };
+
+    let [client, httpBackend] = setupTestClient();
+
+    beforeEach(() => {
+        [client, httpBackend] = setupTestClient();
```
Comment (duxovni) [inline spec/integ/matrix-client-syncing.spec.ts:63]: And here

**Review by duxovni — CHANGES_REQUESTED** [review]
The `expect`s that can be compressed into one line are just a nitpick, mostly focused on the un-cleaned-up `httpBackend`s and the `==`s.

File: spec/integ/matrix-client-methods.spec.ts [hunk 13]
```diff
@@ -13,68 +13,84 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-            httpBackend.when(
+            httpBackend!.when(
                 "POST", "/_matrix/media/r0/upload",
             ).check(function(req) {
                 expect(req.rawData).toEqual(buf);
-                expect(req.queryParams.filename).toEqual("hi.txt");
-                if (!(req.queryParams.access_token == accessToken ||
+                expect(req.queryParams?.filename).toEqual("hi.txt");
+                if (!(req.queryParams?.access_token == accessToken ||
                         req.headers["Authorization"] == "Bearer " + accessToken)) {
```
Comment (duxovni) [inline spec/integ/matrix-client-methods.spec.ts:75]: ^

**Review by duxovni — APPROVED** [review]
Looks good, there's still the `==` here but it's a minor nitpick in the context of a test.

**Pull request merged**